mmcls/models/retriever/image2image.py

```diff
@@ -0,0 +1,315 @@
+        feats = self.extract_feat(inputs)
+        if isinstance(feats, tuple):
+            feats = feats[-1]
+
+        # Matching of similarity
+        result = self.matching(feats)
+
+        pred_scores = result['score']
+        pred_labels = result['pred_label']
+        if data_samples is not None:
+            for data_sample, score, label in zip(data_samples, pred_scores,
+                                                 pred_labels):
+                data_sample.set_pred_score(score).set_pred_label(label)
+        else:
+            data_samples = []
+            for score, label in zip(pred_scores, pred_labels):
+                data_samples.append(ClsDataSample().set_pred_score(
+                    score).set_pred_label(label))
+        self.post_process(data_samples)
+        return data_samples
+
+    def _get_prototype_from_dataloader(self, device):
+        data_loader = self.prototype
+        num = len(data_loader.dataset)
+
+        self.prototype_vecs = None
+        for data_batch in data_loader:
+            batch_num = len(data_batch['inputs'])
+            data = self.data_preprocessor(data_batch, False)
+            feat = self.extract_feat(data['inputs'])
+            if isinstance(feat, tuple):
+                feat = feat[-1]
+            if self.prototype_vecs is None:
+                dim = feat.shape[-1]
+                self.prototype_vecs = torch.zeros(num, dim).to(device)
+            for i in range(batch_num):
+                sample_idx = data_batch['data_samples'][i].get('sample_idx')
+                self.prototype_vecs[sample_idx] = feat[i]
+
+        rank, world_size = get_dist_info()
+        if world_size > 1:
+            dist.barrier()
+            dist.all_reduce(
+                self.prototype_vecs, dist.ReduceOp.SUM, async_op=True)
+
+    @torch.no_grad()
+    def prepare_prototype(self):
+        """Used in meta testing. This function will be called before the meta
+        testing. Obtain the vector based on the prototype.
+
+        - torch.Tensor: The prototype vector is the prototype
+        - str: The path of the extracted feature path, parse data structure,
+            and generate the prototype feature vector set
+        - Dataloader or config: Extract and save the feature vectors according
+            to the dataloader
+        """
+
+        device = next(self.image_encoder.parameters()).device
+        if isinstance(self.prototype, torch.Tensor):
+            self.prototype_vecs = self.prototype
+            self.prototype_vecs = self.prototype_vecs.to(device)
+        elif isinstance(self.prototype, str):
+            self.prototype_vecs = torch.load(self.prototype).to(device)
+        if isinstance(self.prototype, dict):
+            self.prototype = Runner.build_dataloader(self.prototype)
+        if isinstance(self.prototype, DataLoader):
+            self._get_prototype_from_dataloader(device)
+        self.prototype_inited = True
+
+    def post_process(self, data_samples):
```

Comment:
It may be better as a private fn, as ` _get_predictions(self, data_samples)`

```
    def _get_predictions(self, result, data_samples):
        """Post-process the output of retiever"""
        if self.topk != -1:
            topk = min(self.topk, data_samples[0].pred_label.score.shape[0])
            result['score] = result['score'][:topk])
            result['pred_label] = result['pred_label'][:topk])

        if data_samples is not None:
            for data_sample, score, label in zip(data_samples, pred_scores,
                                                 pred_labels):
                data_sample.set_pred_score(score).set_pred_label(label)
        else:
            data_samples = []
            for score, label in zip(pred_scores, pred_labels):
                data_samples.append(ClsDataSample().set_pred_score(
                    score).set_pred_label(label))

        return data_samples
```

And `predict` can be:
```
    def predict(self,
                inputs: tuple,
                data_samples: Optional[List[ClsDataSample]] = None,
                **kwargs) -> List[ClsDataSample]:
        """
        if not self.prototype_inited:
            self.prepare_prototype()

        feats = self.extract_feat(inputs)
        if isinstance(feats, tuple):
            feats = feats[-1]

        # Matching of similarity
        result = self.matching(feats)

        return self._get_predictions(result, data_samples)
```